components/org.wso2.carbon.identity.branding.preference.resolver/src/main/java/org/wso2/carbon/identity/branding/preference/resolver/UIBrandingPreferenceResolverImpl.java

```diff
@@ -180,6 +182,45 @@ Get ancestor organization ids (including itself) of a given organization. The li
     }
 
     @Override
+    public void clearBrandingResolverCacheHierarchy(String currentTenantDomain) throws BrandingPreferenceMgtException {
+        String organizationId = getOrganizationId();
+
+        try {
+            OrganizationManager organizationManager =
+                        BrandingResolverComponentDataHolder.getInstance().getOrganizationManager();
+
+            if (organizationId == null) {
+                // If organization id is not available in the context, try to resolve it from tenant domain
+                organizationId = organizationManager.resolveOrganizationId(currentTenantDomain);
+            }
+
+            List<String> childOrganizationIds = new ArrayList<>();
+            childOrganizationIds.add(organizationId);
+
+            // Clear branding resolver caches by looping (breadth-first) through child organization hierarchy
+            while (!childOrganizationIds.isEmpty()) {
```

Comment:
As we have small org trees, the implementation is fine to merge. If we consider performance, we can do following improvements.

1 - Fetch the tenant domains of the sub-tree in a breadth first order in one database call. 
Ex - 
```SELECT UM_DOMAIN_NAME, SUB_ORGS.DEPTH FROM UM_TENANT INNER JOIN (SELECT UM_ID, DEPTH FROM UM_ORG_HIERARCHY INNER JOIN UM_TENANT ON UM_ORG_HIERARCHY.UM_PARENT_ID = UM_TENANT.UM_ORG_UUID 
WHERE UM_DOMAIN_NAME=<Branding-applied-tenant-domain>) AS SUB_ORGS ON UM_TENANT.UM_ORG_UUID = SUB_ORGS.UM_ID ORDER BY SUB_ORGS.DEPTH ASC
```

Note:- not tried this query yet.

2 - Offload the work of clear cache if exist to a executor service.

Reply:
These suggestions are noted as future improvements.